extensions/amp-carousel/0.2/amp-carousel.js

```diff
@@ -0,0 +1,160 @@
+import {Carousel} from './carousel.js';
+import {htmlFor} from '../../../src/static-template';
+import {isExperimentOn} from '../../../src/experiments';
+import {isLayoutSizeDefined} from '../../../src/layout';
+
+/**
+ * @param {!Element} el The Element to check.
+ * @return {boolean} Whether or not the Element is a sizer Element.
+ */
+function isSizer(el) {
+  return el.tagName == 'I-AMPHTML-SIZER';
+}
+
+class AmpCarousel extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {number} */
+    this.advanceCount_ = 1;
+
+    /** @private {?Carousel} */
+    this.carousel_ = null;
+
+    /** @private {!Array<!Element>} */
+    this.slides_ = [];
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  buildCallback() {
+    const {element} = this;
+    // Grab the slides up front so we can place them later.
+    this.slides_ = Array.from(element.children).filter(c => !isSizer(c));
+    // Create the "Shadow DOM"
+    element.appendChild(this.createDom_());
+
+    this.carousel_ = new Carousel({
+      win,
+      element,
+      scrollContainer: this.element.querySelector('.i-amphtml-carousel-scroll'),
+      runMutate: cb => this.mutateElement(cb),
+    });
+
+    // Handle the initial set of attributes
+    Array.from(this.element.attributes).forEach(attr => {
+      this.attributeChanged_(attr.name, attr.value);
+    });
+
+    this.setupActions_();
+
+    // Do some manual "slot" distribution
+    const scrollContainer = element.querySelector('.i-amphtml-carousel-scroll');
+    this.slides_.forEach(slide => {
+      slide.classList.add('i-amphtml-carousel-slotted');
+      scrollContainer.appendChild(slide);
+    });
+
+    this.carousel_.updateSlides(this.slides_);
+    // Signal for runtime to check children for layout.
+    return this.mutateElement(() => {});
+  }
+
+  /** @override */
+  isRelayoutNeeded() {
+    return true;
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.carousel_.updateUi();
+  }
+
+  /** @override */
+  mutatedAttributesCallback(mutations) {
+    for (const key in mutations) {
+      this.attributeChanged_(key, mutations[key]);
+    }
+  }
+
+  /**
+   * @private
+   */
+  createDom_() {
+    return htmlFor(this.element)`
+      <div class="i-amphtml-carousel-scroll"></div>
+    `;
+  }
+
+  /**
+   * @private
+   */
+  setupActions_() {
+    this.registerAction('prev', () => this.carousel_.prev(), ActionTrust.LOW);
+    this.registerAction('next', () => this.carousel_.next(), ActionTrust.LOW);
+    this.registerAction('goToSlide', ({args}) => {
+      this.carousel_.goToSlide(args['index'] || -1);
+    }, ActionTrust.LOW);
+  }
+
+  /**
+   * @param {string} name The name of the attribute.
+   * @param {string} newValue The new value of the attribute.
+   * @private
+   */
+  attributeChanged_(name, newValue) {
+    switch (name) {
```

Comment:
I can see this getting long. You could shorten the byte size of this method to `1/3 + k`.

```js
const boolStr = s => s == 'true';
const numOrZero = n => Number(n) || 0;

const attributeUpdateMethodName = attr => 'update' +
  dashToCamelCase(attr).replace(/^[a-z]/, (_, c) => c.toUpperCase());

// ... 

const mutatable = dict({
   'auto-advance': boolStr,
   'auto-advance-count': numOrZero,
   // ...
});

if (hasOwn(mutatable, name)) {
   const parsed = mutatable[name](newValue);
   this.carousel_[attributeUpdateMethodName(name)](parsed);
}
```

It also has the added benefit that it forces the `Carousel` surface to maintain consistency with attributes 1:1.

Reply:
While this works well for regular JavaScript, this won't play nice when using Closure Compiled code, since the prototype method is renamed at compile time. In this case, the prototype methods will likely be removed entirely as the compiler thinks that they are not called.

Reply:
Public methods don't get mangled with our current infrastructure, right?

Either way, not a big deal for now but I can see this growing linearly.

Reply:
Hmm, I'm not sure, but I think if is only used from a single extension, it might still be mangled. If it doesn't currently, I think we might be moving that way in the future.